ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/PipelineChainDownloader.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static tech.pegasys.pantheon.util.FutureUtils.completedExceptionally;
+import static tech.pegasys.pantheon.util.FutureUtils.exceptionallyCompose;
+
+import tech.pegasys.pantheon.ethereum.eth.manager.EthScheduler;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncTarget;
+import tech.pegasys.pantheon.services.pipeline.Pipeline;
+
+import java.util.Optional;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PipelineChainDownloader<C> implements ChainDownloader {
+  private static final Logger LOG = LogManager.getLogger();
+  private final SyncTargetManager<C> syncTargetManager;
+  private final DownloadPipelineFactory downloadPipelineFactory;
+  private final EthScheduler scheduler;
+
+  private final AtomicBoolean cancelled = new AtomicBoolean(false);
+  private Pipeline<?> currentDownloadPipeline;
+
+  public PipelineChainDownloader(
+      final SyncTargetManager<C> syncTargetManager,
+      final DownloadPipelineFactory downloadPipelineFactory,
+      final EthScheduler scheduler) {
+    this.syncTargetManager = syncTargetManager;
+    this.downloadPipelineFactory = downloadPipelineFactory;
+    this.scheduler = scheduler;
+  }
+
+  @Override
+  public CompletableFuture<Void> start() {
+    return performDownload();
```

Comment:
You may want to use an `AtomicBoolean` to make sure start is only called once.

Reply:
We do tend to follow that pattern.  Done.